webapp/src/main/java/life/qbic/events/DomainEventSerializer.java

```diff
@@ -0,0 +1,63 @@
+package life.qbic.events;
+
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import life.qbic.usermanagement.registration.UserRegistered;
+
+public class DomainEventSerializer {
+
+  public <T extends DomainEvent> String serialize(T event) {
+    if (event instanceof UserRegistered) {
+      return UserRegisteredSerializer.serialize((UserRegistered) event);
+    }
+    throw
+        new UnrecognizedEventTypeException(
+        "Cannot serialize events of type " + event.getClass().getName() + ". Unknown event type.");
+  }
+
+  public DomainEvent deserialize(String serializedEvent, String desiredType) {
+    if (desiredType.equals(UserRegistered.class.getName())) {
+      return UserRegisteredSerializer.deserialize(serializedEvent);
+    }
+    throw new UnrecognizedEventTypeException(
+        "Cannot deserialize events of type " + desiredType + ". Unknown event type.");
+  }
+
+  private static class UserRegisteredSerializer {
+
+    static String serialize(UserRegistered event) {
+      HashMap<String, String> propertiesMap = new HashMap<>();
+      propertiesMap.put("userId", event.getUserId());
+      propertiesMap.put("occurredOn", event.occurredOn().toString());
+      return propertiesMap.toString();
+    }
+
+    static UserRegistered deserialize(String serializedEvent) {
+      var cleaned = serializedEvent.replaceAll("\\{", "").replaceAll("}", "");
+      String[] propertyPairs = cleaned.split(", ");
+      String userId = null;
+      Instant occurredOn = null;
+      for (String propertyPair : propertyPairs) {
+        String[] keyValuePair = propertyPair.split("=");
+        String key = keyValuePair[0];
+        String value = keyValuePair[1];
+        if (key.equals("userId")) {
+          userId = value;
+        } else if (key.equals("occurredOn")) {
+          occurredOn = Instant.parse(value);
+        }
+      }
+      if (userId == null) {
+        throw new RuntimeException("Cannot deserialize to type " + UserRegistered.class.getName());
+      } else if (occurredOn == null) {
+        throw new RuntimeException("Cannot deserialize to type " + UserRegistered.class.getName());
+      }
+      UserRegistered deserializedEvent = new UserRegistered(userId);
+      deserializedEvent.setOccurredOn(occurredOn);
+      return deserializedEvent;
+    }
```

Comment:
Why the pain and implement serialisation and de-serialisation? We can use Java's build in Object Serialisation and Deserialisation features. 

For example create an StoredEvent value object, use Javas ObjectOutputStream (https://docs.oracle.com/javase/7/docs/api/java/io/ObjectOutputStream.html) and write it as Blob in the DB. Same works in the reverse order, with ObjectInputStream.

Another issue is, that you did not save the event type. This is important information, since the type determines all later routing, within and for remote contexts. The Domain Event class name as String is sufficient, following the example from Vaughn Vernon.

We also need an event id for de-duplication processing.

Reply:
I suggest to introduce an object serialisation interface, and go with a default implementation (like `StringObjectSerialization`) and later we can introduce JSON, or Javas native Object Serialisation

Reply:
and yes, we need the id (incremented by one) so we preserve event order information. We get that for free from the data base.

Reply:
the event order information is perserved in the stored event :) it does however not have anything to do with the event store

Reply:
how is that?

Reply:
The stored event is assigned a `long id` from the event store. This id does not propagate out of the context of the event store and is used by the event store to handle event order information. Furhter the event order information can be read by comparing occurredOn timestamps.